File: python/fibonacci.py
#!/usr/bin/env python3
"""
Fibonacci - Recursive vs Iterative in Python
With memoization demonstration
"""

import functools
import time


def fib_recursive_naive(n, depth=0):
    """Naive recursive fibonacci - exponential time"""
    indent = "  " * depth
    print(f"{indent}→ fib({n}) called")
    
    if n <= 1:
        print(f"{indent}← fib({n}) = {n} [BASE CASE]")
        return n
    
    result = fib_recursive_naive(n - 1, depth + 1) + fib_recursive_naive(n - 2, depth + 1)
    print(f"{indent}← fib({n}) = {result}")
    return result


def fib_iterative(n):
    """Iterative fibonacci - linear time, constant space"""
    if n <= 1:
        return n
    
    prev2, prev1 = 0, 1
    print(f"fib(0) = 0")
    print(f"fib(1) = 1")
    
    for i in range(2, n + 1):
        current = prev1 + prev2
        print(f"fib({i}) = {prev1} + {prev2} = {current}")
        prev2, prev1 = prev1, current
    
    return prev1


@functools.lru_cache(maxsize=None)
def fib_memoized(n, depth=0):
    """Memoized recursive fibonacci - linear time and space"""
    indent = "  " * depth
    print(f"{indent}→ fib({n}) called")
    
    if n <= 1:
        print(f"{indent}← fib({n}) = {n} [BASE CASE]")
        return n
    
    result = fib_memoized(n - 1, depth + 1) + fib_memoized(n - 2, depth + 1)
    print(f"{indent}← fib({n}) = {result} [CACHED]")
    return result


def fib_generator():
    """Generator - produces infinite fibonacci sequence"""
    a, b = 0, 1
    while True:
        yield a
        a, b = b, a + b


def count_calls_decorator(func):
    """Decorator to count function calls"""
    def wrapper(*args, **kwargs):
        wrapper.calls += 1
        return func(*args, **kwargs)
    wrapper.calls = 0
    return wrapper


@count_calls_decorator
def fib_counted(n):
    """Fibonacci with call counting"""
    if n <= 1:
        return n
    return fib_counted(n - 1) + fib_counted(n - 2)


def benchmark_approaches(n):
    """Compare different fibonacci approaches"""
    print("\n" + "="*60)
    print(f"BENCHMARK: Computing fib({n})")
    print("="*60)
    
    # Iterative
    print("\n1. Iterative approach:")
    start = time.time()
    result_iter = fib_iterative(n) if n <= 10 else None
    elapsed_iter = time.time() - start
    if result_iter:
        print(f"   Result: {result_iter}")
    print(f"   Time: {elapsed_iter:.6f} seconds")
    print(f"   Space: O(1) - constant")
    
    # Naive recursive (only for small n)
    if n <= 10:
        print("\n2. Naive recursive approach:")
        fib_counted.calls = 0
        start = time.time()
        result_naive = fib_counted(n)
        elapsed_naive = time.time() - start
        print(f"   Result: {result_naive}")
        print(f"   Time: {elapsed_naive:.6f} seconds")
        print(f"   Calls: {fib_counted.calls}")
        print(f"   Space: O(n) - stack depth")
    
    # Memoized
    print("\n3. Memoized recursive approach:")
    fib_memoized.cache_clear()  # Clear cache
    start = time.time()
    result_memo = fib_memoized(n) if n <= 10 else None
    elapsed_memo = time.time() - start
    if result_memo:
        print(f"   Result: {result_memo}")
    cache_info = fib_memoized.cache_info()
    print(f"   Time: {elapsed_memo:.6f} seconds")
    print(f"   Cache: hits={cache_info.hits}, misses={cache_info.misses}")
    print(f"   Space: O(n) - cache + stack")
    
    # Generator
    print("\n4. Generator approach:")
    start = time.time()
    fib_gen = fib_generator()
    result_gen = None
    for i, val in enumerate(fib_gen):
        if i == n:
            result_gen = val
            break
    elapsed_gen = time.time() - start
    print(f"   Result: {result_gen}")
    print(f"   Time: {elapsed_gen:.6f} seconds")
    print(f"   Space: O(1) - generator state only")


def demonstrate_list_comprehension():
    """Show Python's functional approach"""
    print("\n" + "="*60)
    print("PYTHON FUNCTIONAL APPROACH")
    print("="*60)
    
    # Generate first 10 fibonacci numbers
    print("\nFirst 10 fibonacci numbers using generator:")
    fib_gen = fib_generator()
    first_10 = [next(fib_gen) for _ in range(10)]
    print(f"  {first_10}")
    
    # Using reduce (fold) - very functional!
    from functools import reduce
    print("\nUsing functools.reduce (like fold in functional languages):")
    print("  Computing fib(10) with reduce:")
    result = reduce(
        lambda acc, _: (acc[1], acc[0] + acc[1]),
        range(10),
        (0, 1)
    )[0]
    print(f"  Result: {result}")


if __name__ == "__main__":
    print("╔═══════════════════════════════════════════════════════════╗")
    print("║   Python Fibonacci - Multiple Approaches & Optimization  ║")
    print("╚═══════════════════════════════════════════════════════════╝")
    
    n = 5
    
    print("\n=== Naive Recursive (with visualization) ===")
    print(f"Computing fib({n}):")
    result = fib_recursive_naive(n)
    print(f"\nFinal result: {result}")
    
    print("\n=== Iterative Approach ===")
    result = fib_iterative(n)
    print(f"Final result: {result}")
    
    print("\n=== Memoized Recursive (much faster!) ===")
    fib_memoized.cache_clear()
    result = fib_memoized(n)
    print(f"\nFinal result: {result}")
    print(f"Cache info: {fib_memoized.cache_info()}")
    
    benchmark_approaches(10)
    demonstrate_list_comprehension()
    
    print("\n" + "="*60)
    print("KEY PYTHON INSIGHTS:")
    print("="*60)
    print("• @lru_cache decorator = automatic memoization!")
    print("• Generators = lazy evaluation, infinite sequences")
    print("• List comprehensions = concise iteration")
    print("• Multiple paradigms: imperative, functional, OO")
    print("• No manual memory management")
    print("• Much more expressive than C/assembly")
    print("="*60)

File: rust/fibonacci.rs
// Fibonacci in Rust - Multiple approaches
// Shows performance, safety, and modern features

use std::time::Instant;
use std::collections::HashMap;

// 1. Naive recursive - exponential time
fn fib_recursive_naive(n: u32, depth: usize) -> u64 {
    let indent = "  ".repeat(depth);
    println!("{}→ fib({}) called", indent, n);
    
    if n <= 1 {
        println!("{}← fib({}) = {} [BASE CASE]", indent, n, n);
        return n as u64;
    }
    
    let result = fib_recursive_naive(n - 1, depth + 1) + 
                 fib_recursive_naive(n - 2, depth + 1);
    println!("{}← fib({}) = {}", indent, n, result);
    result
}

// 2. Iterative - linear time, constant space
fn fib_iterative(n: u32) -> u64 {
    if n <= 1 {
        return n as u64;
    }
    
    let (mut prev2, mut prev1) = (0u64, 1u64);
    println!("fib(0) = 0");
    println!("fib(1) = 1");
    
    for i in 2..=n {
        let current = prev1 + prev2;
        println!("fib({}) = {} + {} = {}", i, prev1, prev2, current);
        prev2 = prev1;
        prev1 = current;
    }
    
    prev1
}

// 3. Memoized with HashMap - dynamic programming
fn fib_memoized_helper(n: u32, memo: &mut HashMap<u32, u64>, depth: usize) -> u64 {
    let indent = "  ".repeat(depth);
    
    if let Some(&result) = memo.get(&n) {
        println!("{}→ fib({}) = {} [CACHED]", indent, n, result);
        return result;
    }
    
    println!("{}→ fib({}) called", indent, n);
    
    if n <= 1 {
        println!("{}← fib({}) = {} [BASE CASE]", indent, n, n);
        memo.insert(n, n as u64);
        return n as u64;
    }
    
    let result = fib_memoized_helper(n - 1, memo, depth + 1) + 
                 fib_memoized_helper(n - 2, memo, depth + 1);
    
    println!("{}← fib({}) = {} [STORING]", indent, n, result);
    memo.insert(n, result);
    result
}

fn fib_memoized(n: u32, visualize: bool) -> u64 {
    let mut memo = HashMap::new();
    let depth = if visualize { 0 } else { 1000 };  // Hide output for large n
    fib_memoized_helper(n, &mut memo, depth)
}

// 4. Const fn - compile-time fibonacci!
const fn fib_const(n: u32) -> u64 {
    if n <= 1 {
        n as u64
    } else {
        // Can't recurse in const fn easily, so use iteration
        let mut prev2 = 0u64;
        let mut prev1 = 1u64;
        let mut i = 2u32;
        
        while i <= n {
            let current = prev1 + prev2;
            prev2 = prev1;
            prev1 = current;
            i += 1;
        }
        prev1
    }
}

// Compile-time constant!
const FIB_10: u64 = fib_const(10);

// 5. Iterator-based approach (lazy evaluation)
struct FibonacciIterator {
    prev: u64,
    curr: u64,
}

impl FibonacciIterator {
    fn new() -> Self {
        FibonacciIterator { prev: 0, curr: 1 }
    }
}

impl Iterator for FibonacciIterator {
    type Item = u64;
    
    fn next(&mut self) -> Option<Self::Item> {
        let result = self.prev;
        let next = self.prev + self.curr;
        self.prev = self.curr;
        self.curr = next;
        Some(result)
    }
}

fn benchmark_approaches(n: u32) {
    println!("\n{}", "=".repeat(60));
    println!("BENCHMARK: Computing fib({})", n);
    println!("{}", "=".repeat(60));
    
    // Iterative
    println!("\n1. Iterative:");
    let start = Instant::now();
    let result = if n <= 10 {
        fib_iterative(n)
    } else {
        let mut prev2 = 0u64;
        let mut prev1 = 1u64;
        for _ in 2..=n {
            let current = prev1 + prev2;
            prev2 = prev1;
            prev1 = current;
        }
        prev1
    };
    let duration = start.elapsed();
    println!("  Result: {}", result);
    println!("  Time: {:?}", duration);
    println!("  Space: O(1) - constant");
    
    // Memoized (no visualization for large n)
    println!("\n2. Memoized:");
    let start = Instant::now();
    let result = fib_memoized(n, false);
    let duration = start.elapsed();
    println!("  Result: {}", result);
    println!("  Time: {:?}", duration);
    println!("  Space: O(n) - memoization table");
    
    // Iterator
    println!("\n3. Iterator:");
    let start = Instant::now();
    let result = FibonacciIterator::new()
        .nth(n as usize)
        .unwrap_or(0);
    let duration = start.elapsed();
    println!("  Result: {}", result);
    println!("  Time: {:?}", duration);
    println!("  Space: O(1) - iterator state only");
    
    // Const (compile-time)
    println!("\n4. Compile-time const:");
    println!("  fib(10) computed at compile time = {}", FIB_10);
    println!("  Zero runtime cost!");
}

fn demonstrate_iterator() {
    println!("\n{}", "=".repeat(60));
    println!("RUST ITERATOR - LAZY EVALUATION");
    println!("{}", "=".repeat(60));
    
    println!("\nFirst 10 Fibonacci numbers:");
    let fibs: Vec<u64> = FibonacciIterator::new()
        .take(10)
        .collect();
    println!("  {:?}", fibs);
    
    println!("\nSum of first 10:");
    let sum: u64 = FibonacciIterator::new()
        .take(10)
        .sum();
    println!("  {}", sum);
    
    println!("\nIterators are:");
    println!("  • Zero-cost abstractions");
    println!("  • Lazily evaluated");
    println!("  • Composable with .map(), .filter(), .take(), etc.");
    println!("  • As fast as hand-written loops!");
}

fn demonstrate_ownership() {
    println!("\n{}", "=".repeat(60));
    println!("OWNERSHIP IN RECURSION");
    println!("{}", "=".repeat(60));
    
    println!("\nRust's ownership prevents common recursion bugs:");
    println!("  • No use-after-free");
    println!("  • No double-free");
    println!("  • No memory leaks (without unsafe)");
    
    println!("\nEach recursive call:");
    println!("  1. Gets its own stack frame");
    println!("  2. Owns its local variables");
    println!("  3. Cleans up automatically on return");
    println!("  4. All checked at compile time!");
}

fn main() {
    println!("╔═══════════════════════════════════════════════════════════╗");
    println!("║    Rust Fibonacci - Performance & Modern Features        ║");
    println!("╚═══════════════════════════════════════════════════════════╝\n");
    
    let n = 5u32;
    
    println!("=== Naive Recursive (with visualization) ===");
    println!("Computing fib({}):", n);
    let result = fib_recursive_naive(n, 0);
    println!("\nFinal result: {}", result);
    
    println!("\n=== Iterative Approach ===");
    let result = fib_iterative(n);
    println!("Final result: {}", result);
    
    println!("\n=== Memoized Recursive ===");
    let result = fib_memoized(n, true);
    println!("\nFinal result: {}", result);
    
    benchmark_approaches(20);
    demonstrate_iterator();
    demonstrate_ownership();
    
    println!("\n{}", "=".repeat(60));
    println!("KEY RUST INSIGHTS:");
    println!("{}", "=".repeat(60));
    println!("• const fn = compile-time computation");
    println!("• Iterators = zero-cost lazy evaluation");
    println!("• No garbage collector needed");
    println!("• Type inference makes code clean");
    println!("• Ownership prevents memory bugs");
    println!("• Performance on par with C/C++");
    println!("• Modern syntax without sacrificing speed");
    println!("{}", "=".repeat(60));
}
